Drand's threat model documentation (#284)
drand/drand: docs/SECURITY.md

description](https://www.researchgate.net/publication/225722958_Secure_Distributed_Key_Generation_for_Discrete-Log_Based_Cryptosystems)
linked is from Gennaro's paper that explains the protocol and its assumptions in
clean way.

**Synchronous Network**: A packet sent from an alive node reaches its
destination in a bounded amount of time. Drand realizes this assumptions by the
usage of timeouts during the DKG protocol.

**Synchronized Clocks**: All nodes must have roughly synchronized clocks. 

**Reliable Broadcast Channel**: When a node broadcasts a packet to all other
nodes, each other node is guaranteed to receive the same exact packet after some
bounded amount of time. This assumption is not strictly realized by drand
currently. See [DKG attacks](#dkg-attacks) section to understand the impact.

**Authenticated Channel**: Every communication between nodes must be
authenticated. Drand achieves this by signing every outgoing DKG packets with
a BLS signature on the longterm public key of the sender node.

**Public Group**: Every nodes willing to run the DKG must know the group
formation before starting the DKG, including the longterm public keys of each

> node. During the DKG, there might be some nods offline or misbehaving. The set

Reviewer: ```suggestion
node. During the DKG, there might be some nodes offline or misbehaving. The set
```

of nodes that _successfully_ passed the DKG are called the _qualified_ set of

> nodes (QUAL). These nodes only have a valid shares and are able to produce

Reviewer: ```suggestion
nodes (QUAL). Only these nodes have valid shares and are able to produce
```

honest at a given time to broadcast their partial signature in order to create
the final random beacon. See the attack section to know what are the
consequences when that is not the case.

**Determinism of the chain**: The chain is deterministic with respect to a fresh
DKG phase. It means, if a trust party gathered a threshold of private shares, it
could generate any beacon of the chain. When a resharing occurs, the individual
shares of each drand node change but the chain remains the same as well. If the
same set of nodes perform a new fresh DKG, it will create a new chain from
scratch.

**Resharing**: During a resharing, a drand network A (with threshold
`tA`)creates new shares for a drand network B (with threshold `tB`) which can be
disjoinct from A, such the drand network B is now responsible to continue
creating drand beacons, while that the distributed public key doesn't change.
For this to happen, there needs to be at least `tA` nodes from network A and
`tB` nodes alive and honest during the resharing. At the end of the protocol,
there is gonna be at least `tB` nodes that are qualified and have private shares
to generate randomness.

## Attacks

> ### Randomness Generation

Reviewer: can we add an analysis of the earliest that a node / external party can learn about the next random value. I don't think front-running (trying to learn future values early) is a direct threat, but it's a concern of what an attacker may be after that's likely relevant to at least some drand consumers.

e.g. are there selective DDOS i could do to make sure my node is the only one that talks to enough others to be able to reconstruct the next round, and then start computation with that early compared to others and give myself an advantage?

Author: Good point, I'll add a section on that.

Reviewer 2: 👍🏽 thank you for bringing that one up. It has came in conversations multiple times :)

There can be multiple ways of attacking the drand network during the randomness
generation phase, each with different consequences.

#### DDoS the drand network

**Scenario**: There is a DDoS attacks on multiple drand nodes and at least a

> threshold of honest drand nodes are now considered offline and can't get other's

Reviewer: what threshold would make the attack effective?

Author: The threshold specified during the group DKG. So it's at least > 50% of the nodes.

the drand randomness stays _unbiasable_: attacker is not able to change the
randomness in any way.

**Mitigation**: Proactive resharing allows both to 
1. Let a new group of nodes take over the randomness generation, potentially
   with more nodes and higher threshold. 
2. Refresh shares for nodes that participate in the resharing in the new group:
   partial beacons created from an old share is not validated by the members of
   the new group.

Given a "periodical" resharing with more nodes, it makes it harder for the
attacker to maintain the grasp on the shares of the drand nodes since he must
have continuous control over the drand node itself. If the operator of a
corrupted drand node recovers from the attacker's intrusion, after a resharing,
the attacker's share is invalid. Moreover, a resharing with more nodes highers
the bar for the attacker to attain the second scenario both because of the
previous argument and because attacker needs now to corrupts more nodes than in
the initial group now.

As such, it is recommended to reshare often, _even if_ between the same nodes,
as it creates new shares.

### Distributed Key Generation

> #### DDoS attacks

Reviewer: this is mitigated by the fact that at the end of the DKG process all nodes know which subset completed the process, and could potentially decide to retry if they aren't happy with that result?

Author: So the DKG fails is there are less than a threshold of nodes stated as "QUALIFIED". If it's not the case, then yes, if some nodes gets dropped, operators can simply re-run the DKG  or re-reshare to retry.
I'll add more details around that.

# Security Model of drand

## Table of Contents

* [Notations](#notations)
* [Model](#model)
   * [Distributed Key Generation security model](#distributed-key-generation-se
-model)
   * [Randomness generation model](#randomness-generation-model)
* [Attacks](#attacks)
   * [Randomness Generation](#randomness-generation)
      * [DDoS the drand network](#ddos-the-drand-network)
      * [Corruption of the drand network](#corruption-of-the-drand-network)
   * [Distributed Key Generation](#distributed-key-generation)
      * [DDoS attacks](#ddos-attacks)
      * [Corruption attacks](#corruption-attacks)
      * [Broadcast Channel Assumption](#broadcast-channel-assumption)

> * [TO REVIEW:](#to-review)

Reviewer 2: Can you generate this using npmjs.com/doctoc so that it is always up to date? (We use it for many other specs)

creation of the randomness. Each drand node have a longterm public key and a
private share (after running the setup/resharing phase). The drand network is
the set of drand nodes connected with each other.

**Relay node**: a node that is connected to a drand daemon and exposing a
Internet-facing interface allowing to fetch the public randomness. The relay
network is the set of relay nodes, partially / potentially connected with each
other.

When the type of the node is not specified in the document, it is assumed from
the context - most often it refers to a drand node.

**Corrupted node**: a node who is in the control of an attacker. In this case
the attacker has access to all the cryptographic material this node posses as
well as the networking authorization. For example, if a relay node is corrupted,
an attacker has a direct connection to a drand node.

**Offline node**: a node who is unreachable from an external point of view. It
can be offline from the point of view of another drand node or a relay node. The
document tries to clarify in which context when relevant.

**Alive node**: a node which is running the binary and sends packets out to the
inernet.

> ## Model

Reviewer 2: ```suggestion
## Security Model
```

In drand, there are two phases which do not require the same security
assumptions. This section highlights both models and the practical realization
or assumptions taken.

> ### Distributed Key Generation security model

Reviewer 2: ```suggestion
### Distributed Key Generation Ceremony (setup phase)
```

Author: So i would like to discuss change of terminology explicitely - I am not sure about the use of "ceremony" for this - 
For trusted setup ceremony, it's ONE TIME THING. Here, DKG happens multiple time over the life of the chain, and it should happen multiple times, even between the same participants to refresh the share. 
Ceremony has this whole context coming from blockchains which I'm not sure we should bear the weight...

Author: Ok you convinced me :)  - then we should use that terminology everywhere then.

The DKG protocol model follows the one from the Pedersen's protocol. The [paper
description](https://www.researchgate.net/publication/225722958_Secure_Distributed_Key_Generation_for_Discrete-Log_Based_Cryptosystems)
linked is from Gennaro's paper that explains the protocol and its assumptions in
clean way.

**Synchronous Network**: A packet sent from an alive node reaches its
destination in a bounded amount of time. Drand realizes this assumptions by the
usage of timeouts during the DKG protocol.

> **Synchronized Clocks**: All nodes must have roughly synchronized clocks. 

Reviewer 2: Roughly? to the second?

honest at a given time to broadcast their partial signature in order to create
the final random beacon. See the attack section to know what are the
consequences when that is not the case.

**Determinism of the chain**: The chain is deterministic with respect to a fresh
DKG phase. It means, if a trust party gathered a threshold of private shares, it
could generate any beacon of the chain. When a resharing occurs, the individual
shares of each drand node change but the chain remains the same as well. If the
same set of nodes perform a new fresh DKG, it will create a new chain from
scratch.

**Resharing**: During a resharing, a drand network A (with threshold
`tA`)creates new shares for a drand network B (with threshold `tB`) which can be
disjoinct from A, such the drand network B is now responsible to continue
creating drand beacons, while that the distributed public key doesn't change.
For this to happen, there needs to be at least `tA` nodes from network A and
`tB` nodes alive and honest during the resharing. At the end of the protocol,
there is gonna be at least `tB` nodes that are qualified and have private shares
to generate randomness.

> ## Attacks

Reviewer 2: ```suggestion
## Threat Model
```

Author: This section is describing concrete attacs, this is different than describing the model under which these attacks occur (like network etc).

Reviewer 2: ```suggestion
## Attack Vectors
```

able to listen to plaintext traffic from the network of a threshold of nodes might be
able to see a threshold number of partial beacons before any other honest nodes.

**Consequence**: The attacker in such position is able to aggregate the final
beacon of the current round before any other drand nodes. The attacker doesn't
have to submit the beacon to the relay node and can already use the beacon for
the application that uses drand. The delay between the time the attacker has the
final beacon and the other nodes have it depends on the network connectivity
between the drand nodes. Under normal circumstances, given a global passive listener
attacker (which is arguably a strong model), the delay between the time the
attacker computes the final beacon and the application's end users receive it
from the relay network is on the order of hundreds of ms: the time that at least
a drand nodes computes the final beacon plus the time it sends it to a relay
node plus the relay network distribute it to the end-users. Applications using
drand should be using the round number as a marker and not the time accuracy
which may not be granular enough for some applications.

> 

Reviewer 2: @nikkolasg you kind of hint at it with your last sentences, but could you add an explicit "Mitigation" section in which you highlight that for use a Drand round as an Epoch, given that each round happens every X seconds (30 seconds, correct?) the advantage of having the value a few hundred ms earlier is non existent.

> #### DDoS the drand network

Reviewer 2: ```suggestion
#### DDoS the drand private network (the drand node network)
```

Author: I explicitely use the notation that the "drand network" is the network of drand nodes. Is it really necessary here again ?

Author: in the notation section

Reviewer 3: This should just be DoS and not DDoS. The attack is also realisable by BGP hijacking or tripping on power cables (though with different degree of difficulty/realism and different mitigation strategies).

   partial beacons created from an old share is not validated by the members of
   the new group.

Given a "periodical" resharing with more nodes, it makes it harder for the
attacker to maintain the grasp on the shares of the drand nodes since he must
have continuous control over the drand node itself. If the operator of a
corrupted drand node recovers from the attacker's intrusion, after a resharing,
the attacker's share is invalid. Moreover, a resharing with more nodes highers
the bar for the attacker to attain the second scenario both because of the
previous argument and because attacker needs now to corrupts more nodes than in
the initial group now.

As such, it is recommended to reshare often, _even if_ between the same nodes,
as it creates new shares.

### Distributed Key Generation

#### DDoS attacks

If during the DKG, some nodes are DDoS attacked, then these nodes might not be
able to receive the deals (shares) in time and / or reply in the second phase in
time. Given the necessity of time for achieving the synchronous network
assumption, that means these nodes risk getting excluded from the final group
that gets shares at the end.

> Practical Remediation: At the end of a DKG, the nodes that successfully ran the

Reviewer 2: ```suggestion
**Practical Remediation**: At the end of a DKG, the nodes that successfully ran the
```

chain advances if conditions are there (enough partial beacon and time for a new round).

**Synchronized Clocks**: All nodes must have roughly synchronized clocks to
start the rounds at the same time. The accuracy of the synchronicity between
clocks only needs to be at the order the round frequency (order of tens of
seconds), which is much higher than the reality of the server's clock (NTP-synced
servers achieve [offsets of under a second over the globe](https://www.eecis.udel.edu/~mills/database/reports/ntp-survey99-minar.pdf)). 

**Broadcast channel**: The randomness generation models only needs a regular
broadcast channel. It does **not** need to be reliable given the deterministic
nature of a partial beacon and the drand chain: for a given round, a given node
can only send one valid partial beacon.

**Threshold**: The threshold is the amount of nodes that must be online and
honest at a given time to broadcast their partial signature in order to create
the final random beacon. See the attack section to know what are the
consequences when that is not the case.

**Determinism of the chain**: The chain is deterministic with respect to a fresh

> DKG phase. It means, if a trust party gathered a threshold of private shares, it
> could generate any beacon of the chain. When a resharing occurs, the individual

Reviewer: ```suggestion
DKG phase. This means that if an actor collects over a threshold of private shares, it
can generate all future chain beacons. When a resharing occurs, the individual
```

to generate randomness.

## Attacks

### Randomness Generation

There can be multiple ways of attacking the drand network during the randomness
generation phase, each with different consequences.

#### Front Running 

**Scenario**: An attacker that is able to listen passively on the traffic
between nodes (if TLS is not used - which is not a recommened setup) OR that is
able to listen to plaintext traffic from the network of a threshold of nodes might be
able to see a threshold number of partial beacons before any other honest nodes.

**Consequence**: The attacker in such position is able to aggregate the final
beacon of the current round before any other drand nodes. The attacker doesn't
have to submit the beacon to the relay node and can already use the beacon for
the application that uses drand. The delay between the time the attacker has the
final beacon and the other nodes have it depends on the network connectivity
between the drand nodes. Under normal circumstances, given a global passive listener
attacker (which is arguably a strong model), the delay between the time the
attacker computes the final beacon and the application's end users receive it

> from the relay network is on the order of hundreds of ms: the time that at least

Reviewer: hundreds of ms is a lot! i'd hope for 10's

Author: I know i shouldn't have put a concrete number :D 
TBH I have no idea of what this number could be, it really depends on the connectivity of the drand nodes, to the drand relay and connectivity of the enduser to the drand relay, and also depends on the capacity of the attacker to be at every outgoing edges of every node... 
I just thought of the worse number possible but in practice under regular network environment, it could be much lower than that. 

I'll rephrase I think without giving specific number.

Reviewer: Sounds good. i think talking in terms of wide-area round-trip-times when it's purely racing, and then alternatively of how many edges would need to be deactivated for the attacker to learn the value without the  group getting to the public value is maybe the easiest way to frame it.

## Notations

**Drand node**: a node that is running the drand daemon and participating to the
creation of the randomness. Each drand node have a longterm public key and a
private share (after running the setup/resharing phase). The drand network is
the set of drand nodes connected with each other.

**Relay node**: a node that is connected to a drand daemon and exposing a
Internet-facing interface allowing to fetch the public randomness. The relay
network is the set of relay nodes, partially / potentially connected with each
other.

When the type of the node is not specified in the document, it is assumed from
the context - most often it refers to a drand node.

**Corrupted node**: a node who is in the control of an attacker. In this case
the attacker has access to all the cryptographic material this node posses as
well as the networking authorization. For example, if a relay node is corrupted,
an attacker has a direct connection to a drand node.

**Offline node**: a node who is unreachable from an external point of view. It
can be offline from the point of view of another drand node or a relay node. The
document tries to clarify in which context when relevant.

> **Alive node**: a node which is running the binary (drand or relay depending on

Author: Right, it's better. I'll change.

authenticated. Drand achieves this by signing every outgoing DKG packets with
a BLS signature on the longterm public key of the sender node.

**Public Group**: Every nodes willing to run the DKG must know the group
formation before starting the DKG, including the longterm public keys of each
node. During the DKG, there might be some nodes offline or misbehaving. The set
of nodes that _successfully_ passed the DKG are called the _qualified_ set of
nodes (QUAL). Only these nodes have valid shares and are able to produce
partial beacons that can be successfully validated with respect to the
distributed public key.

**DKG's biasability in signatures**: Pedersen's DKG is known to exhibit a
weakness in the biasability in the distribution of the distributed private key.
However, the same authors (Gennaro et al.) that proved the latter also proved
this bias is not relevant in the setting of using the DKG to perform digital
signatures, which offers other strong properties -
[paper](https://pdfs.semanticscholar.org/642b/d1bbc86c7750cef9fa770e9e4ba86bd49eb9.pdf).
In particular, the paper mentions discrete log based systems. However, it is not

> yet strictly proven that system using computation co-CDH assumptions as required

Author: Accepting but will change later - the assumption is more complex than computational diffie hellman.

for threshold BLS signatures are secure in the model of Gennaro.  However, it is
believed that this assumptions holds in this context as well and is being worked
on.  Note that using threshold BLS signatures as a source of randomness _is_
formally proven secured in this [paper](https://eprint.iacr.org/2020/096.pdf)
from Galindo et al. 

### Randomness generation model

**Network**: The randomness generation protocol do not make any assumption on
the network bounds. As soon as packet comes in, node processes them and the
chain advances if conditions are there (enough partial beacon and time for a new round).

**Synchronized Clocks**: All nodes must have roughly synchronized clocks to

> start the rounds at the same time. The precision of the synchronicity between

Reviewer 3: ```suggestion
start the rounds at the same time. The accuracy of the synchronicity between
```

clocks only needs to be at the order the round frequency (order of tens of

> seconds), which is much higher than the reality of server's clock (NTP-synced
> servers have drifts of under a second over the globe). XXX need source.

Reviewer 3: ```suggestion
seconds), which is much higher than the reality of the server's clock (NTP-synced
servers achieve [offsets of under a second over the globe](https://www.eecis.udel.edu/~mills/database/reports/ntp-survey99-minar.pdf)). 
```
This is a fairly standard source, even though pretty old. It's a pretty uncontroversial statement, though -- NTP can generally provide 10s of ms anywhere around the world because there are local (low-latency) stratum 1 servers in every country and in most networks.

can only send one valid partial beacon.

**Threshold**: The threshold is the amount of nodes that must be online and
honest at a given time to broadcast their partial signature in order to create
the final random beacon. See the attack section to know what are the
consequences when that is not the case.

**Determinism of the chain**: The chain is deterministic with respect to a fresh
DKG phase. It means, if a trust party gathered a threshold of private shares, it
could generate any beacon of the chain. When a resharing occurs, the individual
shares of each drand node change but the chain remains the same as well. If the
same set of nodes perform a new fresh DKG, it will create a new chain from
scratch.

**Resharing**: During a resharing, a drand network A (with threshold
`tA`)creates new shares for a drand network B (with threshold `tB`) which can be
disjoinct from A, such the drand network B is now responsible to continue
creating drand beacons, while that the distributed public key doesn't change.
For this to happen, there needs to be at least `tA` nodes from network A and
`tB` nodes alive and honest during the resharing. At the end of the protocol,

> there is gonna be at least `tB` nodes that are qualified and have private shares

Reviewer 3: ```suggestion
there are going to be at least `tB` nodes that are qualified and have private shares
```

different polynomials for example - see
[here](https://www.jcraige.com/vss-forgery) for one example of such an attack.
Note attacker could try to partition the set of honest nodes in two such that
each half would have consistent shares within itself but inconsistent with
respect to the other half.

**Consequence**: Shares can be inconsistent with each other, and in such cases,
nodes will not be able to verify partial beacons during the randomness
generation phase. Another more subtil scenario is that nodes finish the DKG
with half of the honest nodes having a distributed public key different than the
other half, a "split".

**Practical Observation**: After a DKG is setup, nodes (1) publish the
distributed public key they have and (2) start the randomness generation rounds.
The first step enables any third party to verify the distributed public keys are
the same (it is in fact sufficient to verify a threshold of them have the same).
In (2), the chain will not be able to advance and therefore it becomes clear
that the DKG step went wrong.  Given the DKG phase is run once in a while, it is
reasonable to assume nodes can restart the DKG phase in case things have gone
wrong.

> **Remediation to keep assumption true - TODO**: A practical step towards ensuring non

Reviewer 2: ```suggestion
**Remediation to keep assumption true - (not implemented yet)**: A practical step towards ensuring non
```

> equivocation during the DKG phase is to move to a gossiping approach.  Indeed,

Reviewer 2: ```suggestion
equivocation during the DKG phase is to move to a libp2p pubsub overlay.  Indeed,
```
